---
title: "Mendelian Randomization Tutorial: CRP, IL-6, and CHD"
author: "Tayyaba Alvi"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
library(knitr)
library(TwoSampleMR)
library(ieugwasr)
library(dplyr)
library(VariantAnnotation)
library(MRInstruments)
library(gwasglue)
library(ggplot2)
```

# Introduction

This tutorial demonstrates how to perform a Mendelian Randomization (MR) analysis using multiple data sources. We explore the causal relationship between C-reactive protein (CRP) and coronary heart disease (CHD), followed by a deeper look at interleukin-6 (IL-6), a potential upstream regulator of CRP. The results highlight the contrast between observational and MR-based evidence and suggest that IL-6 may play a more direct causal role in CHD risk.

# 1. Accessing GWAS Data from Different Sources

## 1.1 From Local VCF Files Using `gwasvcf_to_TwoSampleMR`
```{r}
vcf_crp <- VariantAnnotation::readVcf("data/ieugwas/CRP_ebi-a-GCST005067_sub.vcf.gz")
crp_exposure <- gwasvcf_to_TwoSampleMR(vcf_crp, "exposure")
crp_exposure <- crp_exposure %>% filter(pval.exposure < 5*10^-8)

vcf_chd <- VariantAnnotation::readVcf("data/ieugwas/CHD_ebi-a-GCST000998_sub.vcf.gz")
chd_outcome <- gwasvcf_to_TwoSampleMR(vcf_chd, "outcome")

harm_data <- harmonise_data(exposure_dat = crp_exposure, outcome_dat = chd_outcome)
```

## 1.2 From MRBase API Using `MRInstruments`
```{r}
vignette("MRBase")
data(gwas_catalog)
exposure_gwas <- subset(gwas_catalog, grepl("GCST007615", STUDY.ACCESSION) &
                          Phenotype_simple == "C-reactive protein levels")
exposure_gwas <- exposure_gwas[exposure_gwas$pval < 5*10^-8, ]
exposure_data <- format_data(exposure_gwas)
head(exposure_data)

# Check available outcomes
ao <- available_outcomes()
head(ao)

# Extract CHD outcome data
chd_outcome_api <- extract_outcome_data(
  snps = exposure_data$SNP,
  outcomes = 'ieu-a-7'
)
```

## 1.3 From Local TSV Files Using `read_exposure_data`
```{r}
exposure <- read_exposure_data(
  filename = "data/gwas_catalogue/CRP-GCST005067_sub.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
exposure$exposure <- "CRP"
exposure <- exposure %>% filter(pval.exposure < 5*10^-8)

outcome <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/CHD_GCST000998.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome$outcome <- "CHD"

# Harmonize and clump
```{r}
dat <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome)
clump <- ld_clump(
  dplyr::tibble(rsid=dat$SNP, pval=dat$pval.exposure, id=dat$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)

harm_clump <- merge(dat, clump, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "id")) %>% unique()
```

# 2. MR Analysis of CRP on CHD
```{r}
res <- mr(harm_clump, method_list = c("mr_egger_regression", "mr_ivw"))
res

mr_heterogeneity(harm_clump)
mr_pleiotropy_test(harm_clump)

# Plots
p1 <- mr_scatter_plot(res, harm_clump)
p1

res_single <- mr_singlesnp(harm_clump)
p2 <- mr_forest_plot(res_single)
p2

res_loo <- mr_leaveoneout(harm_clump)
p3 <- mr_leaveoneout_plot(res_loo)
p3

# Generate report
mr_report(harm_clump, output_path = "figure/")
```

# 3. MR Analysis of CRP on IL-6 (Upstream Relationship)
```{r}
outcome_il6 <- read_outcome_data(
  snps = exposure$SNP,
  filename = "data/gwas_catalogue/IL6-GCST004446.tsv.gz",
  sep = "\t",
  snp_col = "variant_id",
  beta_col = "beta",
  se_col = "standard_error",
  effect_allele_col = "effect_allele",
  other_allele_col = "other_allele",
  eaf_col = "effect_allele_frequency",
  pval_col = "p_value"
)
outcome_il6$outcome <- "IL-6"

il6_data <- harmonise_data(exposure_dat = exposure, outcome_dat = outcome_il6)
clump_il6 <- ld_clump(
  dplyr::tibble(rsid=il6_data$SNP, pval=il6_data$pval.exposure, id=il6_data$exposure),
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p = 0.99,
  pop = "EUR",
  opengwas_jwt = get_opengwas_jwt(),
  bfile = NULL,
  plink_bin = NULL
)
harm_clump_il6 <- merge(il6_data, clump_il6, by.x = c("SNP", "pval.exposure", "exposure"), by.y = c("rsid", "pval", "id")) %>% unique()

res_il6 <- mr(harm_clump_il6, method_list = c("mr_egger_regression", "mr_ivw"))
res_il6

mr_heterogeneity(harm_clump_il6)
mr_pleiotropy_test(harm_clump_il6)

p1_il6 <- mr_scatter_plot(res_il6, harm_clump_il6)
p1_il6

res_single_il6 <- mr_singlesnp(harm_clump_il6)
p2_il6 <- mr_forest_plot(res_single_il6)
p2_il6

res_loo_il6 <- mr_leaveoneout(harm_clump_il6)
p3_il6 <- mr_leaveoneout_plot(res_loo_il6)
p3_il6

mr_report(harm_clump_il6, output_path = "figure/")
```

# Conclusion

This tutorial demonstrated how to conduct MR using multiple data formats (VCF, TSV, API) and data harmonization pipelines. The CRP–CHD analysis showed no evidence of a causal effect, while IL-6—upstream of CRP—demonstrated a potential causal effect on CHD, supporting the hypothesis that IL-6 may be a better intervention target.

